pkg/obs/nodeserver.go

```diff
@@ -17,38 +17,147 @@ limitations under the License.
+	if err := nodePublishValidation(capability, volumeID, targetPath); err != nil {
+		return nil, err
+	}
+
+	credentials := ns.Driver.cloud
+	volume, err := services.GetParallelFSBucket(credentials, volumeID)
+	if err != nil {
+		return nil, err
+	}
+	log.Infof("NodePublishVolume: volume detail: %s", protosanitizer.StripSecrets(volume))
+
+	notMnt, err := ns.Mount.IsLikelyNotMountPointAttach(targetPath)
+	if err != nil {
+		return nil, err
+	}
+	if !notMnt {
+		log.Infof("NodePublishVolume: %s has already mounted", targetPath)
+		return &csi.NodePublishVolumeResponse{}, nil
+	}
+	if err := ns.Mount.MakeDir(targetPath); err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to make dir: %s, error: %v", targetPath, err)
+	}
+	if err := createCredentialFile(ns.Driver.cloud.Global.AccessKey, ns.Driver.cloud.Global.SecretKey); err != nil {
+		return nil, err
+	}
+
+	mntCmd := fmt.Sprintf("obsfs %s %s -o url=obs.%s.%s -o passwd_file=%s %s", volume.BucketName, targetPath,
+		ns.Driver.cloud.Global.Region, ns.Driver.cloud.Global.Cloud, credential, defaultOpts)
+	log.Infof("NodePublishVolume: obsfs cmd : %s", mntCmd)
+
+	if err := sendCommand(mntCmd, "http://unix/mount", ns.MountClient); err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to mount %s at %s: %v",
+			volume.BucketName, targetPath, err)
+	}
+	log.Infof("NodePublishVolume: mount %s at %s successfully", volume.BucketName, targetPath)
+	return &csi.NodePublishVolumeResponse{}, nil
+}
+
+func createCredentialFile(accessKey, secretKey string) error {
+	if err := os.MkdirAll(path.Dir(credentialFile), os.ModePerm); err != nil {
+		return status.Errorf(codes.Internal, "Failed to make dir: %s, error: %v", credentialFile, err)
+	}
+
+	writer, err := os.OpenFile(filepath.Clean(credentialFile), os.O_CREATE|os.O_WRONLY|os.O_TRUNC, perm)
+	if err != nil {
+		return status.Errorf(codes.Internal, "Failed to open file: %s, error: %v", credentialFile, err)
+	}
+	defer writer.Close()
+
+	credentialInfo := accessKey + ":" + secretKey
+	_, err = fmt.Fprintln(writer, credentialInfo)
+	return err
+}
+
+func nodePublishValidation(capability *csi.VolumeCapability, volumeID string, targetPath string) error {
+	if capability == nil {
+		return status.Error(codes.InvalidArgument, "Validation failed, volume capability cannot be nil")
+	}
+	if len(volumeID) == 0 {
+		return status.Error(codes.InvalidArgument, "Validation failed, volumeID cannot be empty")
+	}
+	if len(targetPath) == 0 {
+		return status.Error(codes.InvalidArgument, "Validation failed, targetPath cannot be empty")
+	}
+	return nil
 }
 
 func (ns *nodeServer) NodeUnpublishVolume(_ context.Context, req *csi.NodeUnpublishVolumeRequest) (
 	*csi.NodeUnpublishVolumeResponse, error) {
-	return nil, status.Error(codes.Unimplemented, "")
+	log.Infof("NodeUnpublishVolume: called with args %v", protosanitizer.StripSecrets(*req))
+	volumeID := req.GetVolumeId()
+	targetPath := req.GetTargetPath()
+	if len(volumeID) == 0 {
```

Comment:
It seems that the non-null check on the `volumeID`, and it is not called by other ways, such as calling the API. 
suggest to call the cloud API to check whether the resource exists.

Reply:
done